File: cpp/src/table/Reconstruct.cpp
#include "table/Reconstruct.hpp"

#include <algorithm>
#include <map>

#include "util/Log.hpp"
#include "util/Utf8.hpp"

namespace ctkm::table {
namespace {

constexpr const char* kLogger = "ctkm.table.reconstruct";

std::string trim(const std::string& text) {
    const char* whitespace = " \t\r\n\f\v";
    const auto begin = text.find_first_not_of(whitespace);
    if (begin == std::string::npos) {
        return std::string();
    }
    const auto end = text.find_last_not_of(whitespace);
    return text.substr(begin, end - begin + 1);
}

}  // namespace

TableCell::TableCell(int cellRow, int cellCol, std::string cellText,
                     std::optional<BoundingBox> cellBox, int cellRowSpan, int cellColSpan)
    : row(cellRow),
      col(cellCol),
      rowSpan(std::max(1, cellRowSpan)),
      colSpan(std::max(1, cellColSpan)),
      box(cellBox),
      text(trim(cellText)) {}

bool TableCell::isEmpty() const { return trim(text).empty(); }

bool TableCell::covers(int targetRow, int targetCol) const {
    return row <= targetRow && targetRow < row + rowSpan && col <= targetCol &&
           targetCol < col + colSpan;
}

Table::Table(std::vector<TableCell> cells, int rows, int cols, std::string source,
             double confidence)
    : cells_(std::move(cells)),
      nRows_(rows),
      nCols_(cols),
      source_(std::move(source)),
      confidence_(confidence) {
    // Giống ``__post_init__`` bản Python: mở rộng kích thước lưới theo ô thực tế.
    for (const auto& cell : cells_) {
        nRows_ = std::max(nRows_, cell.row + cell.rowSpan);
        nCols_ = std::max(nCols_, cell.col + cell.colSpan);
    }
}

bool Table::isEmpty() const {
    for (const auto& cell : cells_) {
        if (!cell.isEmpty()) {
            return false;
        }
    }
    return true;
}

const TableCell* Table::cellAt(int row, int col) const {
    for (const auto& cell : cells_) {
        if (cell.covers(row, col)) {
            return &cell;
        }
    }
    return nullptr;
}

std::vector<const TableCell*> Table::rowCells(int row) const {
    std::vector<const TableCell*> found;
    for (const auto& cell : cells_) {
        if (cell.row <= row && row < cell.row + cell.rowSpan) {
            found.push_back(&cell);
        }
    }
    std::stable_sort(found.begin(), found.end(),
                     [](const TableCell* a, const TableCell* b) { return a->col < b->col; });
    return found;
}

std::vector<const TableCell*> Table::columnCells(int col) const {
    std::vector<const TableCell*> found;
    for (const auto& cell : cells_) {
        if (cell.col <= col && col < cell.col + cell.colSpan) {
            found.push_back(&cell);
        }
    }
    std::stable_sort(found.begin(), found.end(),
                     [](const TableCell* a, const TableCell* b) { return a->row < b->row; });
    return found;
}

std::vector<const TableCell*> Table::sortedCells() const {
    std::vector<const TableCell*> found;
    found.reserve(cells_.size());
    for (const auto& cell : cells_) {
        found.push_back(&cell);
    }
    std::stable_sort(found.begin(), found.end(), [](const TableCell* a, const TableCell* b) {
        if (a->row != b->row) {
            return a->row < b->row;
        }
        return a->col < b->col;
    });
    return found;
}

std::vector<std::vector<std::string>> Table::toGrid() const {
    std::vector<std::vector<std::string>> grid(
        static_cast<std::size_t>(std::max(0, nRows_)),
        std::vector<std::string>(static_cast<std::size_t>(std::max(0, nCols_))));
    for (const auto& cell : cells_) {
        for (int row = cell.row; row < cell.row + cell.rowSpan; ++row) {
            for (int col = cell.col; col < cell.col + cell.colSpan; ++col) {
                if (row >= 0 && row < nRows_ && col >= 0 && col < nCols_) {
                    grid[static_cast<std::size_t>(row)][static_cast<std::size_t>(col)] =
                        cell.text;
                }
            }
        }
    }
    return grid;
}

std::string Table::render(std::size_t maxWidth) const {
    const auto grid = toGrid();
    if (grid.empty()) {
        return "(bảng rỗng)";
    }
    std::vector<std::size_t> widths(static_cast<std::size_t>(std::max(0, nCols_)), 0);
    for (const auto& row : grid) {
        for (std::size_t index = 0; index < row.size(); ++index) {
            widths[index] = std::min(maxWidth, std::max(widths[index], utf8::length(row[index])));
        }
    }
    std::string separator = "+";
    for (const auto width : widths) {
        separator += std::string(width + 2, '-');
        separator += '+';
    }

    std::string out = separator;
    for (const auto& row : grid) {
        out += '\n';
        out += '|';
        for (std::size_t index = 0; index < row.size(); ++index) {
            std::string shown = row[index];
            const std::size_t length = utf8::length(shown);
            if (length > widths[index]) {
                shown = utf8::substr(shown, 0, widths[index] - 1) + "…";
            }
            const std::size_t padding = widths[index] - std::min(widths[index], utf8::length(shown));
            out += ' ';
            out += shown;
            out += std::string(padding, ' ');
            out += ' ';
            out += '|';
        }
        out += '\n';
        out += separator;
    }
    return out;
}

std::string joinTokens(const std::vector<OCRToken>& tokens) {
    const auto lines = ocr::groupTokensIntoLines(tokens);
    std::string out;
    for (const auto& line : lines) {
        std::string part;
        for (const auto& token : line) {
            if (token.text.empty()) {
                continue;
            }
            if (!part.empty()) {
                part += ' ';
            }
            part += token.text;
        }
        if (part.empty()) {
            continue;
        }
        if (!out.empty()) {
            out += ' ';
        }
        out += part;
    }
    return trim(out);
}

int assignTokensToCells(std::vector<TableCell>& cells, const std::vector<OCRToken>& tokens,
                        double minOverlap) {
    std::vector<TableCell*> boxed;
    for (auto& cell : cells) {
        if (cell.box.has_value()) {
            boxed.push_back(&cell);
        }
    }

    int assigned = 0;
    for (const auto& token : tokens) {
        if (token.isEmpty()) {
            continue;
        }
        TableCell* best = nullptr;
        double bestScore = 0.0;
        for (auto* cell : boxed) {
            const double score = token.box.overlapRatio(*cell->box);
            if (score > bestScore) {
                bestScore = score;
                best = cell;
            }
        }
        if (best == nullptr || bestScore < minOverlap) {
            log::debug(kLogger, "Token '" + token.text + "' không khớp ô nào");
            continue;
        }
        best->tokens.push_back(token);
        ++assigned;
    }

    for (auto* cell : boxed) {
        if (!cell->tokens.empty()) {
            cell->text = joinTokens(cell->tokens);
        }
    }
    return assigned;
}

std::vector<std::pair<double, double>> mergeIntervals(
    const std::vector<std::pair<double, double>>& intervals, double minGap) {
    if (intervals.empty()) {
        return {};
    }
    std::vector<std::pair<double, double>> ordered = intervals;
    std::stable_sort(ordered.begin(), ordered.end(),
                     [](const std::pair<double, double>& a, const std::pair<double, double>& b) {
                         return a.first < b.first;
                     });
    std::vector<std::pair<double, double>> merged{ordered.front()};
    for (std::size_t index = 1; index < ordered.size(); ++index) {
        auto& current = merged.back();
        if (ordered[index].first <= current.second + minGap) {
            current.second = std::max(current.second, ordered[index].second);
        } else {
            merged.push_back(ordered[index]);
        }
    }
    return merged;
}

std::vector<std::pair<double, double>> detectColumnBands(const std::vector<OCRToken>& tokens) {
    if (tokens.empty()) {
        return {};
    }
    double xMin = tokens.front().box.x1();
    double xMax = tokens.front().box.x2();
    for (const auto& token : tokens) {
        xMin = std::min(xMin, token.box.x1());
        xMax = std::max(xMax, token.box.x2());
    }
    const double tableWidth = std::max(1.0, xMax - xMin);

    // Token quá rộng (tiêu đề, ô merge ngang) làm nhoè biên cột nên bị loại ra
    // khi dựng band, nhưng vẫn được gán vào cột ở bước sau.
    std::vector<OCRToken> narrow;
    for (const auto& token : tokens) {
        if (token.box.width() <= tableWidth * kWideTokenRatio) {
            narrow.push_back(token);
        }
    }
    const std::vector<OCRToken>& source = narrow.empty() ? tokens : narrow;

    std::vector<double> heights;
    heights.reserve(source.size());
    for (const auto& token : source) {
        heights.push_back(token.box.height());
    }
    std::sort(heights.begin(), heights.end());
    const double medianHeight = heights.empty() ? 10.0 : heights[heights.size() / 2];
    // Khoảng trắng giữa 2 cột thường lớn hơn khoảng cách giữa 2 từ cùng dòng.
    const double minGap = std::max(6.0, medianHeight * 0.9);

    std::vector<std::pair<double, double>> intervals;
    intervals.reserve(source.size());
    for (const auto& token : source) {
        intervals.emplace_back(token.box.x1(), token.box.x2());
    }
    return mergeIntervals(intervals, minGap);
}

int bestBand(const BoundingBox& box, const std::vector<std::pair<double, double>>& bands) {
    int bestIndex = 0;
    double bestOverlap = -1.0;
    for (std::size_t index = 0; index < bands.size(); ++index) {
        const double overlap =
            std::min(box.x2(), bands[index].second) - std::max(box.x1(), bands[index].first);
        if (overlap > bestOverlap) {
            bestOverlap = overlap;
            bestIndex = static_cast<int>(index);
        }
    }
    return bestIndex;
}

std::optional<Table> clusterTokensToTable(const std::vector<OCRToken>& tokens,
                                          double lineOverlapThreshold, int maxColumns) {
    std::vector<OCRToken> usable;
    for (const auto& token : tokens) {
        if (!token.isEmpty()) {
            usable.push_back(token);
        }
    }
    if (usable.size() < 2) {
        log::warn(kLogger, "Không đủ token để cluster thành bảng (" +
                               std::to_string(usable.size()) + " token)");
        return std::nullopt;
    }

    const auto rows = ocr::groupTokensIntoLines(usable, lineOverlapThreshold);
    if (rows.size() < 2) {
        log::warn(kLogger,
                  "Chỉ gom được " + std::to_string(rows.size()) + " hàng, không dựng được bảng");
        return std::nullopt;
    }

    auto bands = detectColumnBands(usable);
    if (bands.empty()) {
        log::warn(kLogger, "Không xác định được biên cột từ bounding box");
        return std::nullopt;
    }
    if (static_cast<int>(bands.size()) > maxColumns) {
        log::warn(kLogger, "Phát hiện " + std::to_string(bands.size()) + " cột (>" +
                               std::to_string(maxColumns) +
                               ") - có thể do OCR tách từ quá vụn, vẫn tiếp tục");
        bands.resize(static_cast<std::size_t>(maxColumns));
    }

    std::map<std::pair<int, int>, std::vector<OCRToken>> grouped;
    for (std::size_t rowIndex = 0; rowIndex < rows.size(); ++rowIndex) {
        for (const auto& token : rows[rowIndex]) {
            const int colIndex = bestBand(token.box, bands);
            grouped[{static_cast<int>(rowIndex), colIndex}].push_back(token);
        }
    }

    std::vector<TableCell> cells;
    for (const auto& entry : grouped) {
        BoundingBox box = entry.second.front().box;
        for (std::size_t index = 1; index < entry.second.size(); ++index) {
            box = box.merge(entry.second[index].box);
        }
        TableCell cell(entry.first.first, entry.first.second, joinTokens(entry.second), box);
        cell.tokens = entry.second;
        cells.push_back(std::move(cell));
    }
    if (cells.empty()) {
        return std::nullopt;
    }

    Table table(std::move(cells), static_cast<int>(rows.size()),
                static_cast<int>(bands.size()), "cluster", 0.5);
    log::info(kLogger, "Dựng bảng bằng cluster: " + std::to_string(table.rowCount()) +
                           " hàng x " + std::to_string(table.columnCount()) + " cột");
    return table;
}

Table tableFromRows(const std::vector<std::vector<std::string>>& rows,
                    const std::string& source) {
    std::vector<TableCell> cells;
    std::size_t maxColumns = 0;
    for (std::size_t rowIndex = 0; rowIndex < rows.size(); ++rowIndex) {
        maxColumns = std::max(maxColumns, rows[rowIndex].size());
        for (std::size_t colIndex = 0; colIndex < rows[rowIndex].size(); ++colIndex) {
            cells.emplace_back(static_cast<int>(rowIndex), static_cast<int>(colIndex),
                               rows[rowIndex][colIndex]);
        }
    }
    return Table(std::move(cells), static_cast<int>(rows.size()), static_cast<int>(maxColumns),
                 source, 1.0);
}

}  // namespace ctkm::table

File: python/ctkm_extractor/table/reconstruct.py
"""Mô hình dữ liệu bảng + fallback dựng bảng bằng cách cluster bounding box.

Đây là tầng fallback thứ 2 (sau PP-Structure): gom token thành hàng theo trục y,
rồi gom thành cột theo trục x. Cách này đơn giản, không cần model, nhưng dễ sai
với ô merge dọc - vì vậy PP-Structure vẫn được ưu tiên.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Iterable, Iterator, Sequence

from ..ocr.base import BoundingBox, OCRToken, group_tokens_into_lines

logger = logging.getLogger(__name__)

#: Token có bề rộng lớn hơn tỉ lệ này so với bảng thì không dùng để dựng biên cột
#: (thường là tiêu đề bảng hoặc ô merge ngang).
WIDE_TOKEN_RATIO = 0.45

#: Ngưỡng tối thiểu phần diện tích token nằm trong ô để coi là thuộc ô đó.
MIN_CELL_OVERLAP = 0.30


@dataclass
class TableCell:
    """Một ô của bảng, có hỗ trợ ô gộp (``row_span``/``col_span``)."""

    row: int
    col: int
    row_span: int = 1
    col_span: int = 1
    box: BoundingBox | None = None
    text: str = ""
    tokens: list[OCRToken] = field(default_factory=list)

    def __post_init__(self) -> None:
        self.row_span = max(1, int(self.row_span))
        self.col_span = max(1, int(self.col_span))
        self.text = (self.text or "").strip()

    @property
    def is_empty(self) -> bool:
        return not self.text.strip()

    def occupies(self) -> Iterator[tuple[int, int]]:
        """Sinh toàn bộ toạ độ ``(row, col)`` mà ô này chiếm."""
        for r in range(self.row, self.row + self.row_span):
            for c in range(self.col, self.col + self.col_span):
                yield (r, c)

    def covers(self, row: int, col: int) -> bool:
        return (
            self.row <= row < self.row + self.row_span
            and self.col <= col < self.col + self.col_span
        )


@dataclass
class Table:
    """Bảng đã dựng: danh sách ô + kích thước lưới."""

    cells: list[TableCell] = field(default_factory=list)
    n_rows: int = 0
    n_cols: int = 0
    #: Nguồn dựng bảng: ``pp_structure`` | ``cluster`` | ``manual``.
    source: str = "unknown"
    confidence: float = 0.0

    def __post_init__(self) -> None:
        if self.cells:
            self.n_rows = max(self.n_rows, max(c.row + c.row_span for c in self.cells))
            self.n_cols = max(self.n_cols, max(c.col + c.col_span for c in self.cells))

    @property
    def is_empty(self) -> bool:
        return not any(not cell.is_empty for cell in self.cells)

    def cell_at(self, row: int, col: int) -> TableCell | None:
        """Ô phủ toạ độ ``(row, col)``; ô gộp trả về chính nó ở mọi toạ độ nó phủ."""
        for cell in self.cells:
            if cell.covers(row, col):
                return cell
        return None

    def row_cells(self, row: int) -> list[TableCell]:
        """Các ô thuộc một hàng, sắp theo cột (ô gộp dọc xuất hiện ở mọi hàng nó phủ)."""
        found: list[TableCell] = [c for c in self.cells if c.row <= row < c.row + c.row_span]
        return sorted(found, key=lambda c: c.col)

    def column_cells(self, col: int) -> list[TableCell]:
        """Các ô thuộc một cột, sắp theo hàng."""
        found: list[TableCell] = [c for c in self.cells if c.col <= col < c.col + c.col_span]
        return sorted(found, key=lambda c: c.row)

    def iter_cells(self) -> Iterator[TableCell]:
        yield from sorted(self.cells, key=lambda c: (c.row, c.col))

    def to_grid(self) -> list[list[str]]:
        """Lưới text đầy đủ; ô gộp lặp lại text ở mọi vị trí nó phủ."""
        grid = [["" for _ in range(self.n_cols)] for _ in range(self.n_rows)]
        for cell in self.cells:
            for row, col in cell.occupies():
                if 0 <= row < self.n_rows and 0 <= col < self.n_cols:
                    grid[row][col] = cell.text
        return grid

    def render(self, max_width: int = 40) -> str:
        """Chuỗi dạng bảng ASCII để in khi chạy ``--debug``."""
        grid = self.to_grid()
        if not grid:
            return "(bảng rỗng)"
        widths = [0] * self.n_cols
        for row in grid:
            for index, value in enumerate(row):
                widths[index] = min(max_width, max(widths[index], len(value)))
        lines: list[str] = []
        separator = "+" + "+".join("-" * (w + 2) for w in widths) + "+"
        lines.append(separator)
        for row in grid:
            cells = []
            for index, value in enumerate(row):
                shown = value if len(value) <= widths[index] else value[: widths[index] - 1] + "…"
                cells.append(" " + shown.ljust(widths[index]) + " ")
            lines.append("|" + "|".join(cells) + "|")
            lines.append(separator)
        return "\n".join(lines)


def assign_tokens_to_cells(
    cells: Sequence[TableCell],
    tokens: Iterable[OCRToken],
    min_overlap: float = MIN_CELL_OVERLAP,
) -> int:
    """Gán token OCR vào ô theo diện tích chồng lấn; trả số token gán được.

    Dùng khi cấu trúc ô đến từ PP-Structure (chỉ có toạ độ) còn text đến từ
    VietOCR - hai nguồn được ghép lại bằng toạ độ.
    """
    boxed = [cell for cell in cells if cell.box is not None]
    assigned = 0
    for token in tokens:
        if token.is_empty:
            continue
        best: TableCell | None = None
        best_score = 0.0
        for cell in boxed:
            assert cell.box is not None
            score = token.box.overlap_ratio(cell.box)
            if score > best_score:
                best_score = score
                best = cell
        if best is None or best_score < min_overlap:
            logger.debug("Token %r không khớp ô nào (score=%.2f)", token.text, best_score)
            continue
        best.tokens.append(token)
        assigned += 1

    for cell in boxed:
        if cell.tokens:
            cell.text = join_tokens(cell.tokens)
    return assigned


def join_tokens(tokens: Sequence[OCRToken]) -> str:
    """Ghép text các token trong một ô theo thứ tự đọc, mỗi dòng cách nhau 1 space."""
    lines = group_tokens_into_lines(list(tokens))
    parts = [" ".join(tok.text for tok in line if tok.text) for line in lines]
    return " ".join(part for part in parts if part).strip()


def _merge_intervals(
    intervals: Sequence[tuple[float, float]], min_gap: float
) -> list[tuple[float, float]]:
    """Gộp các khoảng [x1, x2] chồng lấn hoặc cách nhau dưới ``min_gap``."""
    if not intervals:
        return []
    ordered = sorted(intervals, key=lambda i: i[0])
    merged: list[list[float]] = [list(ordered[0])]
    for start, end in ordered[1:]:
        current = merged[-1]
        if start <= current[1] + min_gap:
            current[1] = max(current[1], end)
        else:
            merged.append([start, end])
    return [(start, end) for start, end in merged]


def detect_column_bands(
    tokens: Sequence[OCRToken], min_gap: float | None = None
) -> list[tuple[float, float]]:
    """Suy ra biên các cột từ hình chiếu bounding box lên trục x."""
    if not tokens:
        return []
    x_min = min(t.box.x1 for t in tokens)
    x_max = max(t.box.x2 for t in tokens)
    table_width = max(1.0, x_max - x_min)

    # Token quá rộng (tiêu đề, ô merge ngang) làm nhoè biên cột nên bị loại ra khi
    # dựng band, nhưng vẫn được gán vào cột ở bước sau.
    narrow = [t for t in tokens if t.box.width <= table_width * WIDE_TOKEN_RATIO]
    source = narrow or list(tokens)

    if min_gap is None:
        heights = sorted(t.box.height for t in source)
        median_height = heights[len(heights) // 2] if heights else 10.0
        # Khoảng trắng giữa 2 cột thường lớn hơn khoảng cách giữa 2 từ cùng dòng.
        min_gap = max(6.0, median_height * 0.9)

    intervals = [(t.box.x1, t.box.x2) for t in source]
    return _merge_intervals(intervals, min_gap=min_gap)


def cluster_tokens_to_table(
    tokens: Sequence[OCRToken],
    *,
    line_overlap_threshold: float = 0.4,
    max_columns: int = 24,
) -> Table | None:
    """Dựng bảng bằng cách cluster bounding box: hàng theo y, cột theo x.

    Trả ``None`` nếu không đủ dữ liệu để dựng bảng có nghĩa (dưới 2 hàng hoặc
    không tách được cột), khi đó tầng gọi sẽ fallback sang xử lý raw text.
    """
    usable = [t for t in tokens if not t.is_empty]
    if len(usable) < 2:
        logger.warning("Không đủ token để cluster thành bảng (%d token)", len(usable))
        return None

    rows = group_tokens_into_lines(usable, overlap_threshold=line_overlap_threshold)
    if len(rows) < 2:
        logger.warning("Chỉ gom được %d hàng, không dựng được bảng", len(rows))
        return None

    bands = detect_column_bands(usable)
    if not bands:
        logger.warning("Không xác định được biên cột từ bounding box")
        return None
    if len(bands) > max_columns:
        logger.warning(
            "Phát hiện %d cột (>%d) - có thể do OCR tách từ quá vụn, vẫn tiếp tục",
            len(bands),
            max_columns,
        )
        bands = bands[:max_columns]

    grouped: dict[tuple[int, int], list[OCRToken]] = {}
    for row_index, line in enumerate(rows):
        for token in line:
            col_index = _best_band(token.box, bands)
            grouped.setdefault((row_index, col_index), []).append(token)

    cells: list[TableCell] = []
    for (row_index, col_index), cell_tokens in sorted(grouped.items()):
        box = cell_tokens[0].box
        for token in cell_tokens[1:]:
            box = box.merge(token.box)
        cells.append(
            TableCell(
                row=row_index,
                col=col_index,
                box=box,
                text=join_tokens(cell_tokens),
                tokens=list(cell_tokens),
            )
        )

    if not cells:
        return None

    table = Table(
        cells=cells,
        n_rows=len(rows),
        n_cols=len(bands),
        source="cluster",
        confidence=0.5,
    )
    logger.info("Dựng bảng bằng cluster: %d hàng x %d cột", table.n_rows, table.n_cols)
    return table


def _best_band(box: BoundingBox, bands: Sequence[tuple[float, float]]) -> int:
    """Chỉ số cột có phần chồng lấn theo trục x lớn nhất với ``box``."""
    best_index = 0
    best_overlap = -1.0
    for index, (start, end) in enumerate(bands):
        overlap = min(box.x2, end) - max(box.x1, start)
        if overlap > best_overlap:
            best_overlap = overlap
            best_index = index
    return best_index


def table_from_rows(rows: Sequence[Sequence[str]], source: str = "manual") -> Table:
    """Dựng ``Table`` từ lưới text thuần - tiện cho unit test và fixture."""
    cells: list[TableCell] = []
    for row_index, row in enumerate(rows):
        for col_index, value in enumerate(row):
            cells.append(TableCell(row=row_index, col=col_index, text=str(value)))
    n_cols = max((len(row) for row in rows), default=0)
    return Table(cells=cells, n_rows=len(rows), n_cols=n_cols, source=source, confidence=1.0)
